packages/shield-controller/src/polling-with-timeout-abort.ts

```diff
@@ -0,0 +1,200 @@
+   *
+   * @param requestId - The ID of the request to insert the entry for.
+   * @param timeout - The timeout for the request.
+   * @returns The request entry that was inserted.
+   */
+  #insertRequestEntry(requestId: string, timeout: number) {
+    const abortController = new AbortController();
+
+    // Set a timeout to abort the request if it takes too long
+    const timerId = setTimeout(() => {
+      abortController.abort(this.ABORT_REASON_TIMEOUT);
+    }, timeout);
+
+    // Set the abort handler and listen to the `abort` event
+    const abortHandler = () => {
+      this.#cleanUpOnFinished(requestId);
+    };
+    abortController.signal.addEventListener('abort', abortHandler);
+
+    const requestEntry: RequestEntry = {
+      abortController,
+      abortHandler,
+      timerId,
+    };
+
+    // Insert the request entry
+    this.#requestEntries.set(requestId, requestEntry);
+
+    return requestEntry;
+  }
+
+  /**
+   * Clean up the request entry upon finished (success or failure).
+   * This will remove the abort handler from the AbortSignal, clear the timeout, and remove the request entry.
+   *
+   * @param requestId - The ID of the request to clean up for.
+   * @returns The request entry that was cleaned up, if it exists.
+   */
+  #cleanUpOnFinished(requestId: string): RequestEntry | undefined {
+    const requestEntry = this.#cleanUpRequestEntryIfExists(requestId);
+    if (requestEntry) {
+      requestEntry.abortController.signal.removeEventListener(
+        'abort',
+        requestEntry.abortHandler,
+      );
+    }
+    return requestEntry;
+  }
+
+  /**
+   * Clean up the request entry if it exists.
+   * This will clear the pending timeout, remove the event listener from the AbortSignal, and remove the request entry.
+   *
+   * @param requestId - The ID of the request to handle the abort for.
+   * @returns The request entry that was aborted, if it exists.
+   */
+  #cleanUpRequestEntryIfExists(requestId: string): RequestEntry | undefined {
+    const requestEntry = this.#requestEntries.get(requestId);
+    if (requestEntry) {
+      clearTimeout(requestEntry.timerId); // Clear the timeout
+      this.#requestEntries.delete(requestId); // Remove the request entry
+    }
+    return requestEntry;
+  }
+
+  /**
+   * Delay with an abort signal.
+   * This will delay the execution of the code until the abort signal is triggered.
+   *
+   * @param ms - The number of milliseconds to delay.
+   * @param abortSignal - The abort signal to listen to.
+   * @returns A promise that resolves when the delay is complete.
+   */
+  async #delayWithAbortSignal(ms: number, abortSignal: AbortSignal) {
+    return new Promise((resolve) => {
+      let timer: NodeJS.Timeout | null = null;
+
+      const abortHandlerForDelay = () => {
+        // clear the timeout and resolve the promise
+        // Note: we don't reject the promise as this is only a dummy delay
```

Comment:
This is quite unusual. The convention is that if a signal gets aborted, then the function throws an error, afaik. Why did you decide to do it differently here? I usually suggest to stick to conventions because otherwise people will have a hard time understanding the code.